### ISSUE TITLE
[question] Is the (approximate) Satterthwaite comparable with SAS Satterthwaite?

### ISSUE THREAD

**Issue by hugesingleton** [rvlenth/emmeans#345]
Hello again! :)

In your previous answers I noticed, that sometimes you mention the SAS output, so I guess you have an access to it.
As my future work will involve both SAS and R, I'd like to make sure, if the Satterthwaite offered by emmeans agrees with the one from SAS?

I mean here both the analytical formula and the approximate one. Of course I understand the latter will slightly vary from run to run, but the magnitude of order should be preserved (e.g. 22.35 vs. 22.4 seems still good, but 22.35 vs. 25 does not).

I did some checking for the GLS and found the following agreements with other R routines (but cannot check with SAS yet):

Data will be provided at the end.

Model - GLS:
```
options(contrasts = c("contr.sum", "contr.poly"))
library(nlme)

m <-gls(value ~ time*method,
        correlation =corSymm(form = ~as.numeric(time)|patient_id),
        weights =varIdent(form =~ 1|time), 
        data = x, 
        na.action = na.omit, 
        method = "ML")
```

Agreement between joint_tests("df.error") and anova(type="marginal", adjustSigma=FALSE)
```
> anova(m, type="marginal",adjustSigma = F)
Denom. DF: 72 
            numDF  F-value p-value
(Intercept)     1 357.3905  <.0001
time            2  21.7348  <.0001
method          1   0.0783  0.7804
time:method     2   5.9360  0.0041

> joint_tests(m, mode="df.error")
 model term  df1 df2 F.ratio p.value
 time          2  66  21.735  <.0001
 method        1  66   0.078  0.7805
 time:method   2  66   5.936  0.0043
```

Agreement between joint_tests("asymptotic") and car::Anova(type=3)  (and anova(adjustSigma=TRUE)
```
> car::Anova(m, type=3)
Analysis of Deviance Table (Type III tests)

Response: value
            Df    Chisq Pr(>Chisq)    
(Intercept)  1 357.3905  < 2.2e-16 ***
time         2  43.4697  3.637e-10 ***
method       1   0.0783   0.779589    
time:method  2  11.8719   0.002643 ** 
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

> joint_tests(m, mode="asymptotic")
 model term  df1 df2 F.ratio p.value
 time          2 Inf  21.735  <.0001
 method        1 Inf   0.078  0.7796
 time:method   2 Inf   5.936  0.0026

> anova(m, type="marginal",adjustSigma = T)
Denom. DF: 72 
            numDF  F-value p-value
(Intercept)     1 387.1731  <.0001
time            2  23.5461  <.0001
method          1   0.0848  0.7717
time:method     2   6.4306  0.0027
```

But I couldn't compare Satterthwaite to anything here...:
```
> set.seed(1000)
> joint_tests(m, mode="satt", extra.iter=5)
Analytical Satterthwaite method not available; using appx-satterthwaite
 model term  df1   df2 F.ratio p.value
 time          2 25.74  21.735  <.0001
 method        1 25.93   0.078  0.7818
 time:method   2 25.74   5.936  0.0076
```

... so I simplified the correlation structure to compound symmetry and tried both gls and lme4::lmer with random intercept. Often they happen to be consistent

```
> m_cs <-gls(value ~ time*method,
+         correlation =corCompSymm(form= ~ 1 | patient_id),
+         data = x, 
+         na.action = na.omit, 
+         method = "ML")
> 
> joint_tests(m_cs, mode="satterthwaite")
 model term  df1   df2 F.ratio p.value
 time          2 32.98  38.305  <.0001
 method        1 70.90   0.078  0.7804
 time:method   2 66.79   9.704  0.0002

> joint_tests(m_cs, mode="appx-satterthwaite")
 model term  df1   df2 F.ratio p.value
 time          2 52.25  38.305  <.0001
 method        1 25.96   0.078  0.7818
 time:method   2 52.25   9.704  0.0003

> joint_tests(lme4::lmer(value ~ time * method + (1|patient_id), data=x), mode="satterthwaite")
 model term  df1 df2 F.ratio p.value
 time          2  48  38.305  <.0001
 method        1  24   0.078  0.7820
 time:method   2  48   9.704  0.0003
```
OK, similar, but not identical - and I can do nothing more. To validate, I'd need SAS.

Would it be possible for you to check your implementation of Satterthwaite against SAS one?

PS: I tried with lavasearch2, but I didn't know how to make it working with joint_tests, as it used own method, and not the vcov. I provided:

```
m_cs_lava <- m_cs
lavaSearch2::sCorrect(m_cs_lava) <- TRUE

joint_tests(m_cs_lava, vcov. = lavaSearch2::vcov2(m_cs_lava)[1:6, 1:6]) ???

> lavaSearch2::vcov2(m_cs_lava)    --- where to pass it? (need to remove 2 last rows and columns)
                (Intercept)         time1         time2       method1 time1:method1 time2:method1   corCoef1     sigma2
(Intercept)    2.331860e-02 -7.663084e-19  1.532617e-18  0.000000e+00  0.000000e+00  0.000000e+00 0.00000000 0.00000000
time1         -7.663084e-19  1.517946e-02 -7.589730e-03  0.000000e+00  0.000000e+00  0.000000e+00 0.00000000 0.00000000
time2          1.532617e-18 -7.589730e-03  1.517946e-02  0.000000e+00  0.000000e+00  0.000000e+00 0.00000000 0.00000000
method1        0.000000e+00  0.000000e+00  0.000000e+00  2.331860e-02 -7.663084e-19  1.532617e-18 0.00000000 0.00000000
time1:method1  0.000000e+00  0.000000e+00  0.000000e+00 -7.663084e-19  1.517946e-02 -7.589730e-03 0.00000000 0.00000000
time2:method1  0.000000e+00  0.000000e+00  0.000000e+00  1.532617e-18 -7.589730e-03  1.517946e-02 0.00000000 0.00000000
corCoef1       0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00 0.01604186 0.01220849
sigma2         0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00 0.01220849 0.03712169
```

----

Data:
```
x <- structure(list(patient_id = c(100, 1, 100, 1, 100, 1, 200, 2, 
200, 2, 200, 2, 300, 3, 300, 3, 300, 3, 400, 4, 400, 4, 400, 
4, 500, 5, 500, 5, 500, 5, 600, 6, 600, 6, 600, 6, 700, 7, 700, 
7, 700, 7, 800, 8, 800, 8, 800, 8, 900, 9, 900, 9, 900, 9, 1000, 
10, 1000, 10, 1000, 10, 1100, 11, 1100, 11, 1100, 11, 1200, 12, 
1200, 12, 1200, 12, 1300, 13, 1300, 13, 1300, 13), method = structure(c(1L, 
2L, 1L, 2L, 1L, 2L, 1L, 2L, 1L, 2L, 1L, 2L, 1L, 2L, 1L, 2L, 1L, 
2L, 1L, 2L, 1L, 2L, 1L, 2L, 1L, 2L, 1L, 2L, 1L, 2L, 1L, 2L, 1L, 
2L, 1L, 2L, 1L, 2L, 1L, 2L, 1L, 2L, 1L, 2L, 1L, 2L, 1L, 2L, 1L, 
2L, 1L, 2L, 1L, 2L, 1L, 2L, 1L, 2L, 1L, 2L, 1L, 2L, 1L, 2L, 1L, 
2L, 1L, 2L, 1L, 2L, 1L, 2L, 1L, 2L, 1L, 2L, 1L, 2L), .Label = c("A", 
"B"), class = "factor"), time = structure(c(1L, 1L, 2L, 2L, 3L, 
3L, 1L, 1L, 2L, 2L, 3L, 3L, 1L, 1L, 2L, 2L, 3L, 3L, 1L, 1L, 2L, 
2L, 3L, 3L, 1L, 1L, 2L, 2L, 3L, 3L, 1L, 1L, 2L, 2L, 3L, 3L, 1L, 
1L, 2L, 2L, 3L, 3L, 1L, 1L, 2L, 2L, 3L, 3L, 1L, 1L, 2L, 2L, 3L, 
3L, 1L, 1L, 2L, 2L, 3L, 3L, 1L, 1L, 2L, 2L, 3L, 3L, 1L, 1L, 2L, 
2L, 3L, 3L, 1L, 1L, 2L, 2L, 3L, 3L), .Label = c("Baseline", "M6", 
"M12"), class = "factor"), value = c(2.94666666666667, 2.94666666666667, 
1.28, 2.94666666666667, 3.28, 2.94666666666667, 2.94666666666667, 
2.94666666666667, 2.28, 2.28, 1.28, 1.28, 3.28, 2.61333333333333, 
1.28, 2.28, 1.28, 1.94666666666667, 6.94666666666667, 3.28, 2.28, 
2.94666666666667, 1.94666666666667, 2.61333333333333, 3.94666666666667, 
3.94666666666667, 1.28, 3.61333333333333, 0.613333333333333, 
3.28, 2.94666666666667, 2.94666666666667, 1.94666666666667, 2.94666666666667, 
1.28, 2.61333333333333, 3.94666666666667, 4.61333333333333, 1.61333333333333, 
1.94666666666667, 1.28, 1.94666666666667, 4.28, 2.94666666666667, 
1.61333333333333, 2.61333333333333, 1.28, 2.61333333333333, 7.28, 
3.61333333333333, 1.61333333333333, 2.61333333333333, 1.28, 2.61333333333333, 
4.61333333333333, 2.61333333333333, 4.61333333333333, 1.28, 4.61333333333333, 
1.28, 4.94666666666667, 4.61333333333333, 2.28, 3.28, 1.94666666666667, 
3.28, 4.94666666666667, 4.61333333333333, 3.61333333333333, 4.28, 
3.28, 4.28, 4.94666666666667, 3.28, 1.94666666666667, 2.94666666666667, 
1.94666666666667, 2.61333333333333)), row.names = c(NA, -78L), class = c("tbl_df", 
"tbl", "data.frame"))
```

**Comment by rvlenth** [comment]
I have only access SAS via a virtual desktop and basically never use it any more. The appt-satterthwaitr method is very ad hoc. I would not recommend doing post hoc tests with method = "ML", as it biases the fixed-effect estimates.

Sent from my iPad

On Apr 12, 2022, at 11:21 PM, hugesingleton ***@***.***> wrote:

﻿

Hello again! :)

In your previous answers I noticed, that sometimes you mention the SAS output, so I guess you have an access to it.
As my future work will involve both SAS and R, I'd like to make sure, if the Satterthwaite offered by emmeans agrees with the one from SAS?

I mean here both the analytical formula and the approximate one. Of course I understand the latter will slightly vary from run to run, but the magnitude of order should be preserved (e.g. 22.35 vs. 22.4 seems still good, but 22.35 vs. 25 does not).

I did some checking for the GLS and found the following agreements with other R routines (but cannot check with SAS yet):

Data will be provided at the end.

Model - GLS:

options(contrasts = c("contr.sum", "contr.poly"))
library(nlme)

m <-gls(value ~ time*method,
        correlation =corSymm(form = ~as.numeric(time)|patient_id),
        weights =varIdent(form =~ 1|time),
        data = x,
        na.action = na.omit,
        method = "ML")


Agreement between joint_tests("df.error") and anova(type="marginal", adjustSigma=FALSE)

> anova(m, type="marginal",adjustSigma = F)
Denom. DF: 72
            numDF  F-value p-value
(Intercept)     1 357.3905  <.0001
time            2  21.7348  <.0001
method          1   0.0783  0.7804
time:method     2   5.9360  0.0041

> joint_tests(m, mode="df.error")
 model term  df1 df2 F.ratio p.value
 time          2  66  21.735  <.0001
 method        1  66   0.078  0.7805
 time:method   2  66   5.936  0.0043


Agreement between joint_tests("asymptotic") and car::Anova(type=3) (and anova(adjustSigma=TRUE)

> car::Anova(m, type=3)
Analysis of Deviance Table (Type III tests)

Response: value
            Df    Chisq Pr(>Chisq)
(Intercept)  1 357.3905  < 2.2e-16 ***
time         2  43.4697  3.637e-10 ***
method       1   0.0783   0.779589
time:method  2  11.8719   0.002643 **
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

> joint_tests(m, mode="asymptotic")
 model term  df1 df2 F.ratio p.value
 time          2 Inf  21.735  <.0001
 method        1 Inf   0.078  0.7796
 time:method   2 Inf   5.936  0.0026

> anova(m, type="marginal",adjustSigma = T)
Denom. DF: 72
            numDF  F-value p-value
(Intercept)     1 387.1731  <.0001
time            2  23.5461  <.0001
method          1   0.0848  0.7717
time:method     2   6.4306  0.0027


But I couldn't compare Satterthwaite to anything here...:

> set.seed(1000)
> joint_tests(m, mode="satt", extra.iter=5)
Analytical Satterthwaite method not available; using appx-satterthwaite
 model term  df1   df2 F.ratio p.value
 time          2 25.74  21.735  <.0001
 method        1 25.93   0.078  0.7818
 time:method   2 25.74   5.936  0.0076


... so I simplified the correlation structure to compound symmetry and tried both gls and lme4::lmer with random intercept. Often they happen to be consistent

> m_cs <-gls(value ~ time*method,
+         correlation =corCompSymm(form= ~ 1 | patient_id),
+         data = x,
+         na.action = na.omit,
+         method = "ML")
>
> joint_tests(m_cs, mode="satterthwaite")
 model term  df1   df2 F.ratio p.value
 time          2 32.98  38.305  <.0001
 method        1 70.90   0.078  0.7804
 time:method   2 66.79   9.704  0.0002

> joint_tests(m_cs, mode="appx-satterthwaite")
 model term  df1   df2 F.ratio p.value
 time          2 52.25  38.305  <.0001
 method        1 25.96   0.078  0.7818
 time:method   2 52.25   9.704  0.0003

> joint_tests(lme4::lmer(value ~ time * method + (1|patient_id), data=x), mode="satterthwaite")
 model term  df1 df2 F.ratio p.value
 time          2  48  38.305  <.0001
 method        1  24   0.078  0.7820
 time:method   2  48   9.704  0.0003


OK, similar, but not identical - and I can do nothing more. To validate, I'd need SAS.

Would it be possible for you to check your implementation of Satterthwaite against SAS one?

PS: I tried with lavasearch2, but I didn't know how to make it working with joint_tests, as it used own method, and not the vcov. I provided:

m_cs_lava <- m_cs
lavaSearch2::sCorrect(m_cs_lava) <- TRUE

joint_tests(m_cs_lava, vcov. = lavaSearch2::vcov2(m_cs_lava)[1:6, 1:6]) ???

> lavaSearch2::vcov2(m_cs_lava)    --- where to pass it? (need to remove 2 last rows and columns)
                (Intercept)         time1         time2       method1 time1:method1 time2:method1   corCoef1     sigma2
(Intercept)    2.331860e-02 -7.663084e-19  1.532617e-18  0.000000e+00  0.000000e+00  0.000000e+00 0.00000000 0.00000000
time1         -7.663084e-19  1.517946e-02 -7.589730e-03  0.000000e+00  0.000000e+00  0.000000e+00 0.00000000 0.00000000
time2          1.532617e-18 -7.589730e-03  1.517946e-02  0.000000e+00  0.000000e+00  0.000000e+00 0.00000000 0.00000000
method1        0.000000e+00  0.000000e+00  0.000000e+00  2.331860e-02 -7.663084e-19  1.532617e-18 0.00000000 0.00000000
time1:method1  0.000000e+00  0.000000e+00  0.000000e+00 -7.663084e-19  1.517946e-02 -7.589730e-03 0.00000000 0.00000000
time2:method1  0.000000e+00  0.000000e+00  0.000000e+00  1.532617e-18 -7.589730e-03  1.517946e-02 0.00000000 0.00000000
corCoef1       0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00 0.01604186 0.01220849
sigma2         0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00 0.01220849 0.03712169


________________________________

Data:

x <- structure(list(patient_id = c(100, 1, 100, 1, 100, 1, 200, 2,
200, 2, 200, 2, 300, 3, 300, 3, 300, 3, 400, 4, 400, 4, 400,
4, 500, 5, 500, 5, 500, 5, 600, 6, 600, 6, 600, 6, 700, 7, 700,
7, 700, 7, 800, 8, 800, 8, 800, 8, 900, 9, 900, 9, 900, 9, 1000,
10, 1000, 10, 1000, 10, 1100, 11, 1100, 11, 1100, 11, 1200, 12,
1200, 12, 1200, 12, 1300, 13, 1300, 13, 1300, 13), method = structure(c(1L,
2L, 1L, 2L, 1L, 2L, 1L, 2L, 1L, 2L, 1L, 2L, 1L, 2L, 1L, 2L, 1L,
2L, 1L, 2L, 1L, 2L, 1L, 2L, 1L, 2L, 1L, 2L, 1L, 2L, 1L, 2L, 1L,
2L, 1L, 2L, 1L, 2L, 1L, 2L, 1L, 2L, 1L, 2L, 1L, 2L, 1L, 2L, 1L,
2L, 1L, 2L, 1L, 2L, 1L, 2L, 1L, 2L, 1L, 2L, 1L, 2L, 1L, 2L, 1L,
2L, 1L, 2L, 1L, 2L, 1L, 2L, 1L, 2L, 1L, 2L, 1L, 2L), .Label = c("A",
"B"), class = "factor"), time = structure(c(1L, 1L, 2L, 2L, 3L,
3L, 1L, 1L, 2L, 2L, 3L, 3L, 1L, 1L, 2L, 2L, 3L, 3L, 1L, 1L, 2L,
2L, 3L, 3L, 1L, 1L, 2L, 2L, 3L, 3L, 1L, 1L, 2L, 2L, 3L, 3L, 1L,
1L, 2L, 2L, 3L, 3L, 1L, 1L, 2L, 2L, 3L, 3L, 1L, 1L, 2L, 2L, 3L,
3L, 1L, 1L, 2L, 2L, 3L, 3L, 1L, 1L, 2L, 2L, 3L, 3L, 1L, 1L, 2L,
2L, 3L, 3L, 1L, 1L, 2L, 2L, 3L, 3L), .Label = c("Baseline", "M6",
"M12"), class = "factor"), value = c(2.94666666666667, 2.94666666666667,
1.28, 2.94666666666667, 3.28, 2.94666666666667, 2.94666666666667,
2.94666666666667, 2.28, 2.28, 1.28, 1.28, 3.28, 2.61333333333333,
1.28, 2.28, 1.28, 1.94666666666667, 6.94666666666667, 3.28, 2.28,
2.94666666666667, 1.94666666666667, 2.61333333333333, 3.94666666666667,
3.94666666666667, 1.28, 3.61333333333333, 0.613333333333333,
3.28, 2.94666666666667, 2.94666666666667, 1.94666666666667, 2.94666666666667,
1.28, 2.61333333333333, 3.94666666666667, 4.61333333333333, 1.61333333333333,
1.94666666666667, 1.28, 1.94666666666667, 4.28, 2.94666666666667,
1.61333333333333, 2.61333333333333, 1.28, 2.61333333333333, 7.28,
3.61333333333333, 1.61333333333333, 2.61333333333333, 1.28, 2.61333333333333,
4.61333333333333, 2.61333333333333, 4.61333333333333, 1.28, 4.61333333333333,
1.28, 4.94666666666667, 4.61333333333333, 2.28, 3.28, 1.94666666666667,
3.28, 4.94666666666667, 4.61333333333333, 3.61333333333333, 4.28,
3.28, 4.28, 4.94666666666667, 3.28, 1.94666666666667, 2.94666666666667,
1.94666666666667, 2.61333333333333)), row.names = c(NA, -78L), class = c("tbl_df",
"tbl", "data.frame"))

**Comment by rvlenth** [comment]
I adapted some code from the **lmerTest** package, which involves a numerical Jacobian of the estimated covariance matrix. I checked it for consistency with other results on a few models fitted by other methods. I would not say that was adequate validation in the  regulatory sense.

Getting exactitude in d.f. is really not all that important, and I disagree with "(e.g. 22.35 vs. 22.4 seems still good, but 22.35 vs. 25 does not)." in that really there isn't much difference between 22.35 and 25 d.f. in terms of tail areas. There *is* a big difference between 2.35 and 3.35 though.

Also, for joint tests, I do not do a d.f. calculation for the *set* of linear functions. Internally I only have code for 1-d.f. cases, so when there are more, I use the minimum of those (conservative approach).
